File: Triangle Containment/v2/triangle-containment-2.py
def is_point_inside_triangle(n, big_point, small_point):
    x1, y1 = big_point["point"]
    x2, y2 = small_point["point"]
    def triangle_area(x1, y1, x2, y2, x3, y3):
        return abs((x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)) / 2.0)
    main_triangle_area = triangle_area(0, 0, n, 0, x1, y1)

    sub_triangle1_area = triangle_area(0, 0, n, 0, x2, y2)
    sub_triangle2_area = triangle_area(0, 0, x1, y1, x2, y2)
    sub_triangle3_area = triangle_area(n, 0, x1, y1, x2, y2)

    return main_triangle_area == sub_triangle1_area + sub_triangle2_area + sub_triangle3_area
    
n, right_x_value = map(int, input().split())

left = (0, 0)
right = (right_x_value, 0)

values = []
for i in range(n):
    x, y, weight = map(int, input().split())
    values.append(
        {
            "point": (x, y), 
            "weight": weight,
        }
    )

copy = values[:]
values.sort(key=lambda value: value["point"][1], reverse=True)
for i in range(len(values)):
    total = 0
    current = values[i]
    for j in range(i+1, len(values)):
        next = values[j]
        if is_point_inside_triangle(right_x_value, current, next):
            total += next["weight"]
    values[i]["total"] = total

for value in copy:
    print(value["total"])
        

File: Triangle Containment/v2/triangle-containment-2.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Point {
    long long x, y;
    Point(long long x, long long y) : x(x), y(y) {}
};

struct Value {
    Point point;
    long long weight;
    long long total;
    Value(const Point point, long long weight, long long total) : point(point), weight(weight), total(total) {}
};

bool is_inside_triangle(const Point& p, const Point& p1, const Point& p2, const Point& p3) {
    // Check if point p is inside the triangle formed by p1, p2, and p3 using cross product method.
    long long d1 = (p1.x - p.x) * (p2.y - p.y) - (p2.x - p.x) * (p1.y - p.y);
    long long d2 = (p2.x - p.x) * (p3.y - p.y) - (p3.x - p.x) * (p2.y - p.y);
    long long d3 = (p3.x - p.x) * (p1.y - p.y) - (p1.x - p.x) * (p3.y - p.y);

    return (d1 >= 0 && d2 >= 0 && d3 >= 0) || (d1 <= 0 && d2 <= 0 && d3 <= 0);
}

int main() {
    int n, right_x_value;
    cin >> n >> right_x_value;

    Point left = Point(0, 0);
    Point right = Point(right_x_value, 0);

    vector<Value> values;
    for (int i = 0; i < n; i++) {
        long long x, y, weight;
        cin >> x >> y >> weight;
        values.push_back(Value(Point(x, y), weight, 0));
    }

    for (int i = 0; i < n; i++) {
        long long total = 0;
        Value current = values[i];
        for (int j = 0; j < n; j++) {
            Value next = values[j];
            if (next.point.y > current.point.y ||
                i == j ||
                (current.point.y > 0 && next.point.y < 0) ||
                (current.point.y < 0 && next.point.y > 0)) {
                continue;
            }
            if (is_inside_triangle(right, left, current.point, next.point)) {
                total += next.weight;
            }
        }
        values[i].total = total;
        cout << total << "\n";
    }

    return 0;
}
